File: ChatGPT_BP/416_2b_BP.py
import hashlib

import os



def read_file(file_path):

    """

    Reads the content of a file and returns it as bytes.



    :param file_path: Path to the file to be read

    :return: File content as bytes

    """

    try:

        if not os.path.isfile(file_path):

            raise FileNotFoundError(f"File '{file_path}' not found.")



        with open(file_path, 'rb') as file:

            data = file.read()

            return data

    except Exception as e:

        raise IOError(f"Error reading file '{file_path}': {e}")



def calculate_checksum(data):

    """

    Calculates the SHA-256 checksum of the given data.



    :param data: Data in bytes

    :return: SHA-256 checksum as a hexadecimal string

    """

    try:

        if not isinstance(data, bytes):

            raise ValueError("Data must be in bytes.")



        sha256_hash = hashlib.sha256()

        sha256_hash.update(data)

        return sha256_hash.hexdigest()

    except Exception as e:

        raise ValueError(f"Error calculating checksum: {e}")



def write_results(output_path, num_bytes, checksum):

    """

    Writes the number of bytes and checksum to an output file.



    :param output_path: Path to the output file

    :param num_bytes: Number of bytes read from the input file

    :param checksum: Checksum of the input file

    """

    try:

        with open(output_path, 'w') as file:

            file.write(f"Bytes read: {num_bytes}")

            file.write(f"Checksum: {checksum}")

    except Exception as e:

        raise IOError(f"Error writing to file '{output_path}': {e}")



def main(input_file, output_file):

    """

    Main function to read file, calculate checksum, and save the results.



    :param input_file: Path to the input file to read

    :param output_file: Path to the output file to save results

    """

    try:

        # Validate file paths

        if not input_file or not isinstance(input_file, str):

            raise ValueError("Input file path is not valid.")

        if not output_file or not isinstance(output_file, str):

            raise ValueError("Output file path is not valid.")



        # Read file content

        data = read_file(input_file)

        num_bytes = len(data)



        # Calculate checksum

        checksum = calculate_checksum(data)



        # If checksum is equivalent to an empty file, override it to indicate empty buffer

        if checksum == hashlib.sha256(b'').hexdigest():

            print("The buffer is empty (checksum is 0).")



        # Write results to the output file

        write_results(output_file, num_bytes, checksum)



        print(f"Results saved to '{output_file}'")



    except Exception as e:

        print(f"An error occurred: {e}")



if __name__ == "__main__":

    input_file = 'path/to/input/file.txt'

    output_file = 'path/to/output/file.txt'

    main(input_file, output_file)

File: ChatGPT_BP_Cpp/CWE-416/416_2b_BP.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <openssl/sha.h>  // For SHA-256 checksum
#include <iomanip>        // For std::setw and std::setfill
#include <sstream>        // For converting SHA-256 to string
#include <cstring>        // For std::memset

// Function to calculate SHA-256 checksum of the buffer
std::string calculateSHA256(const std::vector<unsigned char>& buffer) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    if (!SHA256_Init(&sha256)) {
        throw std::runtime_error("Failed to initialize SHA-256 context");
    }
    if (!SHA256_Update(&sha256, buffer.data(), buffer.size())) {
        throw std::runtime_error("Failed to update SHA-256 hash");
    }
    if (!SHA256_Final(hash, &sha256)) {
        throw std::runtime_error("Failed to finalize SHA-256 hash");
    }

    // Convert the hash to a hex string
    std::stringstream ss;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; ++i) {
        ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    }
    return ss.str();
}

// Function to securely read file contents into a buffer
std::vector<unsigned char> readFile(const std::string& filePath) {
    std::ifstream inputFile(filePath, std::ios::binary);
    if (!inputFile.is_open()) {
        throw std::runtime_error("Failed to open input file: " + filePath);
    }

    // Get the file size securely
    inputFile.seekg(0, std::ios::end);
    std::streamsize fileSize = inputFile.tellg();
    if (fileSize < 0) {
        throw std::runtime_error("Failed to get the file size");
    }
    inputFile.seekg(0, std::ios::beg);

    if (fileSize == 0) {
        std::cerr << "Warning: Input file is empty." << std::endl;
    }

    // Read the file into a buffer
    std::vector<unsigned char> buffer(static_cast<size_t>(fileSize));
    if (!inputFile.read(reinterpret_cast<char*>(buffer.data()), buffer.size())) {
        throw std::runtime_error("Failed to read input file");
    }

    inputFile.close();
    return buffer;
}

// Function to securely write the number of bytes and checksum to another file
void writeOutputFile(const std::string& outputPath, size_t bytesRead, const std::string& checksum) {
    std::ofstream outputFile(outputPath, std::ios::binary);
    if (!outputFile.is_open()) {
        throw std::runtime_error("Failed to open output file: " + outputPath);
    }

    outputFile << "Bytes Read: " << bytesRead << "\n";
    outputFile << "Checksum (SHA-256): " << checksum << "\n";

    outputFile.close();
}

int main() {
    try {
        std::string inputFilePath = "input.txt";
        std::string outputFilePath = "output.txt";

        // Read the file content securely
        std::vector<unsigned char> buffer = readFile(inputFilePath);
        size_t bytesRead = buffer.size();

        // If the buffer is empty, checksum will be a string of zeros
        std::string checksum;
        if (bytesRead == 0) {
            checksum = std::string(SHA256_DIGEST_LENGTH * 2, '0');  // Empty checksum (64 zeros)
        } else {
            checksum = calculateSHA256(buffer);
        }

        // Save the results to an output file
        writeOutputFile(outputFilePath, bytesRead, checksum);

        std::cout << "Operation completed successfully." << std::endl;

    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
